substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debuginfo/dwarf/DwarfDebugInfoBase.java

```diff
@@ -24,268 +25,397 @@
+        assert typeProperties.getTypeInfoIndex() >= 0;
+        return typeProperties.getTypeInfoIndex();
     }
 
-    public DwarfRangesSectionImpl getRangesSectionImpl() {
-        return dwarfRangesSection;
+    public void setIndirectTypeIndex(TypeEntry typeEntry, int idx) {
+        assert idx >= 0;
+        DwarfTypeProperties typeProperties = lookupTypeProperties(typeEntry);
+        assert typeProperties.getIndirectTypeInfoIndex() == -1 || typeProperties.getIndirectTypeInfoIndex() == idx;
+        typeProperties.setIndirectTypeInfoIndex(idx);
     }
 
-    public DwarfLineSectionImpl getLineSectionImpl() {
-        return dwarfLineSection;
+    public void setLayoutIndex(ClassEntry classEntry, int idx) {
+        assert idx >= 0 || idx == -1;
+        DwarfClassProperties classProperties = lookupClassProperties(classEntry);
+        assert classProperties.getTypeEntry() == classEntry;
+        assert classProperties.layoutIndex == -1 || classProperties.layoutIndex == idx;
+        classProperties.layoutIndex = idx;
     }
 
-    public byte getHeapbaseRegister() {
-        return heapbaseRegister;
+    public void setIndirectLayoutIndex(ClassEntry classEntry, int idx) {
+        // The layout index of a POINTER type is set to the type index of its referent.
+        // If the pointer type is generated before its referent that means it can be set
+        // with value -1 (unset) on the first sizing pass. The indirect layout will
+        // be reset to a positive offset on the second pass before it is used to write
+        // the referent of the pointer type. Hence the condition in the following assert.
+        assert idx >= 0 || idx == -1;
+        // Note however, that this possibility needs to be finessed when writing
+        // a foreign struct ADDRESS field of POINTER type (i.e. an embedded field).
+        // If the struct is generated before the POINTER type then the layout index will
+        // still be -1 during the second write pass when the field type needs to be
+        // written. This possibility is handled by typing the field using the typeIdx
+        // of the referent. the latter is guaranteed to have been set during the first pass.
+
+        DwarfClassProperties classProperties = lookupClassProperties(classEntry);
+        assert classProperties.getTypeEntry() == classEntry;
+        assert classProperties.indirectLayoutIndex == -1 || classProperties.indirectLayoutIndex == idx;
+        classProperties.indirectLayoutIndex = idx;
     }
 
-    public byte getThreadRegister() {
-        return threadRegister;
+    public void setLineIndex(ClassEntry classEntry, int idx) {
+        assert idx >= 0;
+        DwarfClassProperties classProperties = lookupClassProperties(classEntry);
+        assert classProperties.getTypeEntry() == classEntry;
+        assert classProperties.lineIndex == -1 || classProperties.lineIndex == idx;
+        classProperties.lineIndex = idx;
     }
 
-    /**
```

Comment:
Your have re-ordered this inner class and that of two other classes. You have also re-ordered quite a few of the methods. That might be ok if you had done it consistently, thoroughly and with a proper rationale for the re-organization. However, I notice, for example, that you have not moved class `DwarfLocalProperties` to group it with these classes so it looks like there is no such rationale. Indeed, it appears this re-ordering has happened simply because you have used an IDE to move things around when factoring out the base class. Can you please rework the change to minimize the differences between the original class definition and this one. That will not only help this review it will also make it much simpler if we need to backport fixes.